query-engine/core/src/query_graph_builder/extractors/filters/scalar.rs

```diff
@@ -1,34 +1,58 @@
 use crate::{
     constants::inputs::filters, ParsedInputMap, ParsedInputValue, QueryGraphBuilderError, QueryGraphBuilderResult,
 };
-use connector::{Filter, ScalarCompare, ScalarListCompare};
-use prisma_models::{PrismaValue, ScalarFieldRef};
+use connector::{Filter, JsonCompare, JsonFilterPath, JsonTargetType, ScalarCompare, ScalarListCompare};
+use prisma_models::{PrismaValue, ScalarFieldRef, TypeIdentifier};
 use std::convert::TryInto;
 
-#[tracing::instrument(name = "parse_scalar_field", skip(filter_key, field, input, reverse))]
+#[tracing::instrument(name = "parse_scalar_field", skip(input_map, reverse))]
 pub fn parse(
-    filter_key: &str,
+    mut input_map: ParsedInputMap,
     field: &ScalarFieldRef,
+    reverse: bool,
+) -> QueryGraphBuilderResult<Vec<Filter>> {
+    let json_path: Option<JsonFilterPath> = match input_map.remove(filters::PATH) {
+        Some(v) => Some(parse_json_path(v)?),
+        _ => None,
+    };
+
+    let filters: Vec<Filter> = input_map
+        .into_iter()
+        .map(|(k, v)| parse_impl(&k, v, field, &json_path, reverse))
+        .collect::<QueryGraphBuilderResult<Vec<Vec<_>>>>()?
+        .into_iter()
+        .flatten()
+        .collect();
+
+    if json_path.is_some() && filters.is_empty() {
+        return Err(QueryGraphBuilderError::InputError(
+            "A JSON path cannot be set without a scalar filter.".to_owned(),
+        ));
+    }
+
+    Ok(filters)
+}
+
+fn parse_impl(
+    filter_key: &str,
     input: ParsedInputValue,
+    field: &ScalarFieldRef,
+    json_path: &Option<JsonFilterPath>,
     reverse: bool,
 ) -> QueryGraphBuilderResult<Vec<Filter>> {
-    let filter = match filter_key {
+    if field.type_identifier == TypeIdentifier::Json {
+        return parse_impl_json(filter_key, input, field, json_path, reverse);
```

Comment:
(Subjective) While it is more or less a nitpick, I also think in this case early return is not really clean. If you think about it in abstractions, then the json filter parsing and other filter parsing are on the same abstraction level. Usually, it's a good idea to express the same-ness of the abstraction with similar depth of function calls or dispatcher functions that take care of fanning out the calls.

It's probably better to pull the dispatch up one level to equalize the filters.